mgr/volumes: allow disabling async job threads

rishabh-d-dave opened the pull request:
Add a new mgr config options that allows pausing async purge threads and async cloner threads.

Fixes: https://tracker.ceph.com/issues/61903
Fixes: https://tracker.ceph.com/issues/68630


TODO
- [ ] https://github.com/ceph/ceph/pull/58647#discussion_r1735962486




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

kotreshhr commented on src/pybind/mgr/volumes/fs/volume.py:
```diff
@@ -261,7 +261,8 @@ def remove_subvolume(self, **kwargs):
             with open_volume(self, volname) as fs_handle:
                 with open_group(fs_handle, self.volspec, groupname) as group:
-                    remove_subvol(self.mgr, fs_handle, self.volspec, group, subvolname, force, retainsnaps)
+                    remove_subvol(self.mgr, fs_handle, self.volspec, group,
```
whitespace change, not required as part of this commit?

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4161,6 +4162,44 @@ def test_subvolume_rm_with_snapshots(self):
+        self.assert_sv_is_present_in_trash(sv, sv_dir, sv_tree)
+        time.sleep(60)
+        self.assert_sv_is_present_in_trash(sv, sv_dir, sv_tree)
+
```
This test case needs to extended to validate, if `purge_trash` is set to `true`, the subvolumes shouldbe deleted.

vshankar replied:
This is marked resolved without an explanation. Were the comments incorporated?

rishabh-d-dave replied:
These tests were deleted long ago since we changed the approach on this PR. Therefore, marking this thread as resolved.

kotreshhr commented on src/pybind/mgr/volumes/fs/purge_queue.py:
```diff
@@ -110,4 +110,10 @@ def get_next_job(self, volname, running_jobs):
 
     def execute_job(self, volname, job, should_cancel):
-        purge_trash_entry_for_volume(self.fs_client, self.vc.volspec, volname, job, should_cancel)
+        if not self.vc.mgr.purge_trash:
```
I think the problem here is that the `job` is already added to `self.jobs` (i.e., running_jobs). So if the subvolume is deleted when purge_trash is `False`, the subvolume would never be purged even after `purge_trash` is set `True`  because the `get_next_job` would skip fetching already running_jobs ?

neesingh-rh commented on src/pybind/mgr/volumes/fs/purge_queue.py:
```diff
@@ -110,4 +110,10 @@ def get_next_job(self, volname, running_jobs):
-        purge_trash_entry_for_volume(self.fs_client, self.vc.volspec, volname, job, should_cancel)
+        if not self.vc.mgr.purge_trash:
+            log.info('skipping purging of trash since config option '
+                     '"purge_trash" is true.')
```
It should be `"purge_trash" is false`

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4212,6 +4204,84 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        '''
+        v, sv = self.volname, self._generate_random_subvolume_name()
+        self.run_ceph_cmd(f'fs subvolume create {v} {sv}')
+        self.run_ceph_cmd('config set mgr mgr/volumes/purge_trash false')
```
nit: We should be adding an assert for `mgr/volumes/purge_trash` after using `config get` if we are changing the config option value

vshankar replied:
Why is the check required after setting the config?

neesingh-rh replied:
If we are changing the config value in the tests, then we need to verify by getting it in my opinion.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4212,6 +4204,84 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        x = self.mount_a.run_shell('sudo ls volumes/_deleting').stdout.getvalue().strip()
+        trash_sv_dir = sv_dir.replace('_nogroup', f'_deleting/{x}')
+        trash_sv_dir = trash_sv_dir.replace(sv, '')
+        sv_tree_new = self.mount_a.run_shell(f'sudo tree {trash_sv_dir}').stdout.getvalue()
```
We can use helper method `_note_files_in_sv()` here.

vshankar replied:
@neesingh-rh suggestions isn't taken into consideration, but the comment is marked resolved. @rishabh-d-dave ??

neesingh-rh replied:
@rishabh-d-dave I guess you marked it resolved by mistake.

rishabh-d-dave replied:
Marking this as resolved since we don't use `tree` anymore.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4212,6 +4204,84 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+    def test_subvolume_rm_purge_trash(self):
+        '''
+        Test that when mgr config option mgr/volumes/purge_trash is set to
+        true, running "ceph fs subvolume rm" will move the subvolume to trash
```
s/true/false

rishabh-d-dave replied:
Right, thanks.

rishabh-d-dave commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -475,6 +475,11 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+            'purge_trash',
+            type='bool',
+            default=True,
+            desc='Should files in trash be deleted?'),
```
Done.

vshankar commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -475,6 +475,12 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+            'async_purge_trash',
+            type='bool',
+            default=True,
+            desc='Should subvolumes be asynchronously purged after they have '
```
The description is a question, why?

It should be an explanation, isn't it?

vshankar reviewed (changes requested):
@rishabh-d-dave What about comment https://github.com/ceph/ceph/pull/54396/commits/6b9f3cf1eef5377983d3a119c97b30b72fda5ba6#r1386395240 ?

vshankar commented:
@rishabh-d-dave Is this ready for review?

vshankar commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -489,7 +489,12 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+            'disable_purge_trash',
+            type='bool',
+            default=False,
+            desc='Setting to "true" will disable purging of subvolumes')
```
nit: "Enabling this config disables ..."

vshankar replied:
.. and also note that ongoing purges would backoff.

rishabh-d-dave replied:
Right. Fixed.

vshankar commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -61,24 +61,30 @@ def purge(self, trashpath, should_cancel):
         """
         def rmtree(root_path):
             log.debug("rmtree {0}".format(root_path))
+
```
unrelated line break.

rishabh-d-dave replied:
Can we keep the extra newlines characters I've added? It's easier to read code when block of code are separated not just by indentation but also by newlines.

vshankar replied:
Let's do the in a new PR.

vshankar commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -61,24 +61,30 @@ def purge(self, trashpath, should_cancel):
                 with self.fs.opendir(root_path) as dir_handle:
                     d = self.fs.readdir(dir_handle)
-                    while d and not should_cancel():
+                    while d and should_purge() and not should_cancel():
```
Looks like the way this has been implemented will let the purge threads pickup an entry for synchronisation and then back off as soon as it tries to delete it. I would like to have this for the already processing threads as a way to backoff.

Pausing the purge should be done by using condition variable wait/signal semantics IMO.

Another point about the way `should_purge` lambda is used here - I think it's not required. You could cancel the ongoing deletions using `purge_queue.cance_jobs()`.

rishabh-d-dave replied:
> You could cancel the ongoing deletions using purge_queue.cance_jobs().

For that, we'll need to pass an instance of `purge_queue` to `operations.trash.Trash.purge()` along with `should_purge()`, so that `purge_queue.cancel_jobs()` can be called when `should_purge()` returns true. Is that correct?

vshankar replied:
Not really - you can invoke that from module.py as `self.vc.purge_queue.cancel_jobs(...)`.

rishabh-d-dave replied:
> Not really - you can invoke that from module.py as `self.vc.purge_queue.cancel_jobs(...)`.

Okay, it's having no effect. I removed lambda function `should_purge` and called `cancel_all_jobs()` (which is in `purge_queue.py`) as soon as config opt value is changed. 

Trying to figure now why was subvolume purged even though `cancell_all_jobs()` was called.

rishabh-d-dave replied:
So cancelling all jobs in ineffective and technically buggy.

cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.

Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off. I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.

@vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

vshankar replied:
> So cancelling all jobs in ineffective and technically buggy.
> 
> cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.
> 
> Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off. I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.
> 
> @vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

@rishabh-d-dave based on what we discussed, it this still an issue?

vshankar replied:
> So cancelling all jobs in ineffective and technically buggy.
> 
> cancel_all_jobs() empties the vols queue[[1]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L257) and jobs queue[[2]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L266). But then all volumes that are not in the queue will be added back here[[3]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L49) and execution of cancelled jobs will begin here[[4]](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/volumes/fs/async_job.py#L59). This renders `cancel_all_jobs()` ineffective.

[3] only happens when option `periodic_async_work` is enabled. Furthermore, you'd need to wait on a condition (variable) before that.

> 
> Even if we fix this to make sure that cancelled jobs remain cancelled, we'll need to add some more code to re-enable queuing of cancelled jobs when config opt `disable_purge_trash` is turned off.

This would essentially be:

```
self.q = [e['name'] for e in list_volumes(self.vc.mgr)]
```

> I believe the solution we had on this PR (using lambda function should_purge()) was much simpler.
> 
> @vshankar Let me know if I should proceed to fix `cancel_all_jobs()`.

Is there anything to fix? You'd need to do two things:

- Introduce a condition variable (or use an existing one) on which the threads would wait until a certain condition is valid again.
- invoke cancel_all_jobs() which would effectively drain `self.q`. The threads would backoff and block on the condition variable. If subvolumes are deleted from this time onwards, the respective volume gets added to `self.q`, but the threads are still blocked on the condition variable. When the condition becomes valid, refresh `self.q` - the purge operations should start.

rishabh-d-dave replied:
Yep, I've pushed these changes on this PR some time ago.

vshankar replied:
> Yep, I've pushed these changes on this PR some time ago.

I still see the implementation not using wait-signal semantics - any reason you think its unneeded?

Also, you mentioned about `cancel_all_jobs()` being buggy - is that still the case?

vshankar commented on src/pybind/mgr/volumes/fs/purge_queue.py:
```diff
@@ -106,8 +106,21 @@ def __init__(self, volume_client, tp_size):
         return get_trash_entry_for_volume(self.fs_client, self.vc.volspec, volname, running_jobs)
 
     def execute_job(self, volname, job, should_cancel):
+        if self.fs_client.disable_purge_trash:
```
While this will technically work, I don't think this is the approach this functionality should take. The threads would still be executing over and over again doing nothing, while we would want the threads to be blocked.

rishabh-d-dave replied:
>  threads to be blocked.

IIUC, you want the thread creation to be blocked (after `cancel_all_jobs()` has been run). Is that correct?

vshankar replied:
yes

neesingh-rh commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -518,7 +518,13 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+            'disable_purge_trash',
+            type='bool',
+            default=False,
+            desc='Enabling this config disables purging of subvolumes and '
```
Suggestion: "Disable purging of subvolumes and ongoing purges from trash"

vshankar commented:
@rishabh-d-dave ping?

rishabh-d-dave commented:
> @rishabh-d-dave ping?

Haven't been able to get this due to too many  ongoing PRs. Will get back to it this week, after getting `mgr/vol clone stats` PR ready.

rishabh-d-dave commented:
@vshankar PTAL.

rishabh-d-dave commented:
Rebased on recent main.

vshankar reviewed (changes requested):
@rishabh-d-dave Do we really need 77df6392ac09a430bf24c13e7a1a416a06615c12 if the async threads sleep on a condition variable when the config option to disable purging sub volumes is enabled? Also, this mechanism nicely integrates with disabling clones.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/purge_queue.py:
```diff
@@ -104,10 +104,27 @@ class ThreadPoolPurgeQueueMixin(AsyncJobs):
         return get_trash_entry_for_volume(self.fs_client, self.vc.volspec, volname, running_jobs)
 
     def execute_job(self, volname, job, should_cancel):
+        if self.disable_purge_trash:
```
I'll go through this and get back.

rishabh-d-dave replied:
Have you checked the last commit? `DNM: don't spawn threads when disable_purge_trash is on`. I had kept it separate back then so that it is easy to review new changes.

vshankar replied:
I did and frankly the approach taken here isn't ideal since firstly the threads keep running (described in the comment above) and secondly, since the threads are always running, the DNM commit is trying to not spawn threads.

Instead, if the threads block on a condition variable, then you won't have to do this jugglery around.

rishabh-d-dave replied:
Done.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -3576,14 +3576,6 @@ def test_subvolume_resize_infinite_size_future_writes(self):
         # verify trash dir is clean
         self._wait_for_trash_empty()
 
-    def test_subvolume_rm_force(self):
```
Yes.

rishabh-d-dave commented on qa/tasks/ceph_test_case.py:
```diff
@@ -52,7 +52,7 @@ def get_ceph_cmd_stdout(self, *args, **kwargs):
             kwargs['args'] = args
         kwargs['stdout'] = kwargs.pop('stdout', StringIO())
-        return self.run_ceph_cmd(**kwargs).stdout.getvalue()
+        return self.run_ceph_cmd(**kwargs).stdout.getvalue().strip()
```
Removed as it the decision was against having `.strip()`.

rishabh-d-dave commented:
@vshankar PTAL.

PS: I've used thread.event instead of threading.condition since former despite of being simpler was sufficing and IMO more suitable for the change we want on this PR.

anthonyeleven commented on PendingReleaseNotes:
```diff
@@ -12,6 +12,8 @@
   output is changed from 'STATUS' to 'STATE'. The state of a group snapshot
   that was shown as 'ok' is now shown as 'complete', which is more descriptive.
+* CephFS: It is now possible to pause the threads that asynchronously purge
+  deleted subvolumes using config option "mgr/volumes/pause_purging_trash".
```
by using the config

anthonyeleven commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -896,6 +896,12 @@ Configure the maximum number of concurrent clone operations. The default is 4:
 
    ceph config set mgr mgr/volumes/max_concurrent_clones <value>
 
+Pause the threads that asynchronously purge the trashed subvolumes:
```
s/purge the/purge/

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+    def test_when_changed_to_false_unpurged_subvol_is_purged(self):
+        '''
+        Test that when MGR config option mgr/volumes/pause_purging_trash is
+        is set to false, trased but unpurged subvolume is purged (fully).
```
trased?

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+        for filename in sv_files:
+            self.assertNotIn(filename, sv_files_new)
+
+    def test_when_false_subvol_is_trashed_and_also_purged(self):
```
This is the default behaviour and should be covered with existing tests, yes?

vshankar replied:
What about this comment?

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+        time.sleep(7)
+        self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
+
+        self.run_ceph_cmd(f'config set mgr {self.conf_opt} true')
```
Why set the config again?

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+
+        self.run_ceph_cmd(f'config set mgr {self.conf_opt} true')
+
+    def wait_for_trashcan_to_be_empty(self):
```
I believe there is already a helper for this in test_volumes.

vshankar replied:
What about this comment?

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+        self.run_ceph_cmd(f'config set mgr {self.conf_opt} false')
+        self.wait_for_trashcan_to_be_empty()
+
+    def test_the_default_value(self):
```
Does this really needs to be a test?

rishabh-d-dave replied:
Yes. If this gets messed up, subvolumes won't be purged by default which will qualify as a bug introduced by this PR.

vshankar replied:
Messed up how? You mean some code that gets merged? (that code would be vetted during review)

rishabh-d-dave replied:
> that code would be vetted during review

Yeah but it's possible to miss and one small test won't hurt.   :)

vshankar replied:
That does not make it a candidate for a test.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4219,6 +4211,46 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
         self._wait_for_trash_empty()
 
+
+class TestSubVolumeRm(TestVolumesHelper):
```
Having this commit here helps me easily and quickly test if the code here has had a negative impact on these tests.

vshankar replied:
I don't think that checking for subvolume remove warrants a separate test class.

rishabh-d-dave replied:
It makes it possible to run all tests related to subvolume removal in one go. This particularly helpful when modifying corresponding source code and testing it.

vshankar replied:
Sorry, I don't find that a pressing reason to move it into a separate test class.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+        for filename in sv_files:
+            self.assertNotIn(filename, sv_files_new)
+
+    def test_when_false_subvol_is_trashed_and_also_purged(self):
```
Moving subvolume from volume directory to trash directory is "trashing". Deleting it from trash directory as well is "purging". So, this test case ensures that when `pause_purging_trash` is `false`, subvolume is moved to trash directory as well as deleted from trash directory. Hence, `test_when_false_subvol_is_trashed_and_also_purged`.

neesingh-rh replied:
> This test name is not making much sense.

This is a default case , don't think it requires a separate test case. Anyways, the name is also a bit confusing. We can have it like:`test_subvol_when_pause_purging_trash_is_disabled` or something like it.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
         self._wait_for_trash_empty()
 
 
+class TestPausePurgingTrash(TestVolumesHelper):
```
Already done. I am using `safe_while` wherever possible.

rishabh-d-dave commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -564,6 +570,8 @@ def config_notify(self):
                     elif opt['name'] == "snapshot_clone_no_wait":
                         self.vc.cloner.reconfigure_reject_clones(self.snapshot_clone_no_wait)
+                    elif opt['name'] == "pause_purging_trash":
+                        self.vc.purge_queue.pause_purging(self.pause_purging_trash)
```
Much better, done! :)

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/purge_queue.py:
```diff
@@ -106,8 +110,22 @@ def __init__(self, volume_client, tp_size):
+        # let purge threads run by default.
+        self.should_purge.set()
+
+    def pause_purging(self, pause_purging_trash):
```
Done.

rishabh-d-dave commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -896,6 +896,12 @@ Configure the maximum number of concurrent clone operations. The default is 4:
 
    ceph config set mgr mgr/volumes/max_concurrent_clones <value>
 
+Pause the threads that asynchronously purge the trashed subvolumes:
```
Done.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -308,6 +309,7 @@ def _do_subvolume_io(self, subvolume, subvolume_group=None, create_dir=None,
             self.mount_a.run_shell_payload(f"mkdir -p {io_path}")
 
         log.debug("filling subvolume {0} with {1} files each {2}MB size under directory {3}".format(subvolume, number_of_files, file_size, io_path))
+
```
Strange, I thought this I had removed this. Removed again.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -13,6 +13,7 @@
 
 from tasks.cephfs.cephfs_test_case import CephFSTestCase
 from tasks.cephfs.fuse_mount import FuseMount
+
```
Done.

rishabh-d-dave commented on qa/tasks/cephfs/mount.py:
```diff
@@ -777,7 +777,7 @@ def run_shell(self, args, **kwargs):
     def get_shell_stdout(self, args, timeout=300, **kwargs):
         return self.run_shell(args=args, timeout=timeout, **kwargs).stdout.\
-            getvalue().strip()
+            getvalue()
```
My bad. Commit message needs an update.

rishabh-d-dave replied:
Done, marking this as resolved.

rishabh-d-dave commented:
@vshankar As I had mentioned in standup, I've pushed updated this PR. This PR branch has 8 commits. First 6 commits are same as what you've reviewed last time (besides changes you requested during last review). The 8th commit contains the changes that were requested by you here [1]. And the 7th commit contains necessary changes to files in first 6 commits, these changes eventually needs to squashed into first 6 commits. 

Please check the 8th commit and let me know if it is moving in the right direction. I was seeing some race conditions in the tests added in this PR due to this commit but most of it has been fixed. If 8th commit looks okay, I'll proceed to squash 7th commit and add another commit to disable cloner threads as well (which we had discussed last time) and also update tests, docs and release notes accordingly.

CC @ceph/cephfs 

[1] https://github.com/ceph/ceph/pull/54396#discussion_r1741452794

rishabh-d-dave commented:
@vshankar As I had mentioned in standup, I've pushed updated this PR. This PR branch has 8 commits. First 6 commits are same as what you've reviewed last time (besides changes you requested during last review). The 8th commit contains the changes that were requested by you here [1]. And the 7th commit contains necessary changes to files in first 6 commits, these changes eventually needs to squashed into first 6 commits. 

Please check the 8th commit and let me know if it is moving in the right direction. I was seeing some race conditions in the tests added in this PR due to this commit but most of it has been fixed. If 8th commit looks okay, I'll proceed to squash 7th commit and add another commit to disable cloner threads as well (which we had discussed last time) and also update tests, docs and release notes accordingly.

CC @ceph/cephfs 

[1] https://github.com/ceph/ceph/pull/54396#discussion_r1741452794

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -165,10 +192,33 @@ def run(self):
+        log.info('pause() saved jobs, cancelling jobs now...')
+        self.cancel_all_jobs()
+        log.info('pause() all jobs cancelled')
+        self.threads.clear()
```
This is off-putting. Why do we need to clear the threads list? Once ongoing jobs are canceled, all threads should block (wait) on a condition variable (events in your case).

rishabh-d-dave replied:
Removed.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -42,6 +44,15 @@ def run(self):
                             self.async_job.threads.remove(self)
                             return
+
+                        if not self.async_job.run_event.is_set():
```
This is not making any sense. Why have this check here and terminate the thread, but also wait() on the event in L#69?

rishabh-d-dave replied:
Changed.

rishabh-d-dave commented:
Hey, @vshankar . I've modified the 8th commit as per your last review and made some more change as well. IMO, it's perfect not. Can you please review this commit and confirm that is looks okay? In that case I'll begin squashing some of the commits and proceed to write code for pausing cloner threads.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -35,24 +37,39 @@ def run(self):
                             log.info("thread [{0}] terminating due to reconfigure".format(thread_name))
                             self.async_job.threads.remove(self)
                             return
+
```
unrelated change.

rishabh-d-dave replied:
Sure, I'll remove it.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -35,24 +37,39 @@ def run(self):
                                 self.async_job.jobs[m] = []
                                 self.async_job.q.append(m)
+
+                        if not self.async_job.run_event.is_set():
```
Why is this required?

rishabh-d-dave replied:
This is located in a `while` loop that'll run infinitely until a job is found. And since this loop runs without acquiring a lock. Since we must to wait without acquiring the lock, this code is breaking out of the loop.

rishabh-d-dave replied:
Otherwise thread won't stop until this iteration is finished which means `get_job()`, `register_job()`, `execute_job()` and `unregister_job()` will be run unnecessarily.

rishabh-d-dave replied:
Marking this comment resolved since it was discussed and changes have moved forward since.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -35,24 +37,39 @@ def run(self):
                             break
                         self.async_job.cv.wait(timeout=timo)
+
+                    if not self.async_job.run_event.is_set():
```
and this too?

rishabh-d-dave replied:
When user pauses before threads find the job, the `break` we discussed in above previous comment will take care of pausing. If user pauses while job is being executed, `should_cancel` in `purge()` of `operation/trash.py` will take care of pausing.

But what if we user pauses while thread is between getting the job and executing it? `continue` under the `if` takes care of this case.

rishabh-d-dave replied:
Marking this comment resolved since it was discussed and changes have moved forward since.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -165,10 +197,33 @@ def run(self):
                     for i in range(c, self.nr_concurrent_jobs):
-                        self.threads.append(JobThread(self, self.vc, name="{0}.{1}.{2}".format(self.name_pfx, time.time(), i)))
-                        self.threads[-1].start()
+                        if not self.run_event.is_set():
```
I don't understand this? Why is this required and more so why is it breaking out of the thread?

rishabh-d-dave replied:
So that no new threads are spawned when entire async op machinery has been paused.

rishabh-d-dave replied:
As soon as async op machinery is resumed by the user, the jobs will be continue to run and new thread will also be spawned.

rishabh-d-dave replied:
Marking this comment resolved since it was discussed and changes have moved forward since.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -156,6 +182,7 @@ def run(self):
         log.debug("tick thread {} starting".format(self.name))
         with lock_timeout_log(self.lock):
             while not self.stopping.is_set():
+
```
Removed.

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -165,10 +197,33 @@ def run(self):
+    def pause(self):
+        self.run_event.clear()
+
+        self.saved_q = deepcopy(self.q)
```
I don't get it? Why do this need to make a copy here?

rishabh-d-dave replied:
In `volume.py`, we call `AsyncJobs.queue_job()` to queue the clone or purge job. And when we cancel all the jobs without finishing them in `AsyncJobs.pause()`, all volumes/jobs are removed from `self.q`/`self.jobs`. Thus, `self.q` needs to to re-instated when we resume the worker threads. That's what this code does.

rishabh-d-dave replied:
Marking this comment resolved since it was discussed and changes have moved forward since.

rishabh-d-dave commented:
Had a discussion with Venky about above review comments. I'll post the changes to allow cancelling jobs without removing it from volume queue, remove 2/3 checks from JobThread.run() and remove code to disallow spawning new threads when purging is paused..

rishabh-d-dave commented:
> Had a discussion with Venky about above review comments. I'll post the changes to allow cancelling jobs without removing it from volume queue, remove 2/3 checks from JobThread.run() and remove code to disallow spawning new threads when purging is paused..

The current changes don't break the tests, but with these changes the tests fail. Digging deeper to find the cause and fix it.

vshankar commented:
Spoke to @rishabh-d-dave regarding this. The approach looks fine. Suggest squashing the DNM/TBS commits. I will plan to have a look once done.

rishabh-d-dave commented:
> Spoke to @rishabh-d-dave regarding this. The approach looks fine. Suggest squashing the DNM/TBS commits. I will plan to have a look once done.

Done. @vshankar

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -65,6 +67,12 @@ def rmtree(root_path):
+                        # when purge threads should run and unset/clear purge
+                        # threads should pause. Pausing happens using this
+                        # wait() method.
+                        should_purge.wait()
```
Marked this as resolved since this has been followed through discussions and corresponding changes.

rishabh-d-dave commented:
minor change

rishabh-d-dave commented:
Removed some extra newlines that were left behind after deleting some code changes and removed helper methods that spawn threads since it's only a improvement, unlike before when it was a needed for this feature.

rishabh-d-dave commented:
Rebased to latest main and resolved the merge-conflict.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
> > Spoke to @rishabh-d-dave regarding this. The approach looks fine. Suggest squashing the DNM/TBS commits. I will plan to have a look once done.
> 
> Done. @vshankar
> 
> Let me know if it looks okay. If so, I'll proceed to add some finishing touches to the tests that have been added in this PR and also proceed to add changes for disabling cloner threads (as per our discussion from before on Slack).

Leaving context for review request here again.

@vshankar @ceph/cephfs

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4251,6 +4253,218 @@ def test_subvolume_rm_with_snapshots(self):
+        o = o.lower().strip()
+        self.assertEqual(o, 'false')
+
+    def test_that_subvols_are_purged_by_default(self):
```
What about this comment?

neesingh-rh replied:
Agree with this comment

vshankar commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -263,7 +324,9 @@ def _cancel_jobs(self, volname):
                 self.cancel_cv.wait()
-            self.jobs.pop(volname)
+
+            if update_queue:
```
Have you checked if it suffices to remove the volume from `self.jobs`? I think it should be fine to remove. Please double check.

vshankar replied:
What about this comment @rishabh-d-dave ?

vshankar replied:
ping??

rishabh-d-dave replied:
It doesn't work, because the we would've removed jobs from self.jobs but kept volname in self.q causing get_job() to  crash when it would looks for jobs for a certain volname.

rishabh-d-dave commented:
@v

> > > will plan to have a look once done.
> > 
> > 
> > Done. @vshankar
> > Let me know if it looks okay. If so, I'll proceed to add some finishing touches to the tests that have been added in this PR and also proceed to add changes for disabling cloner threads (as per our discussion from before on Slack).
> 
> Leaving context for review request here again.
> 
> @vshankar @ceph/cephfs

@vshankar Good to proceed to add code for pausing cloner thread?

vshankar commented:
> @vshankar Good to proceed to add code for pausing cloner thread?

Given the nature of changes, I'd say let's keep that as a separate change (and probably have someone else give a shot at implementing it).

neesingh-rh commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -951,6 +951,13 @@ Configure the maximum number of concurrent clone operations. The default is 4:
    ceph config set mgr mgr/volumes/max_concurrent_clones <value>
 
+Pause the threads that asynchronously purge trashed subvolumes. This option is
+useful during cluster recovery scenairos:
```
s/scenairos/scenarios

rishabh-d-dave replied:
Thanks, fixed.

neesingh-rh commented on doc/cephfs/troubleshooting.rst:
```diff
@@ -128,6 +128,13 @@ things to do:
   That prevents any clients from establishing new sessions with the MDS.
 
+* **Turn off async purge threads** The volumes plugin spawns threads for
+  asychronously purging trashed/deleted subvolumes. To help troubleshooting or
```
s/asychronously/asynchronously

rishabh-d-dave replied:
Thanks, fixed.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4337,298 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        '''
+        v = self.volname
+        sv = 'sv1'
+        self.config_set('mgr', self.conf_opt, False)
```
Why are we explicitly setting it to `False` when it is already the default value

neesingh-rh replied:
ping?

rishabh-d-dave replied:
Have you checked the latest code? I don't see this line of code at line number 4566 anymore.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4337,298 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        time.sleep(2)
+
+        # XXX actual test here: test that purging is resumed and finished
+        self.config_set('mgr', self.conf_opt, False)
```
Shouldn't we `self._assert_trashed_sv_has_num_of_files(trashed_sv_path, NUM_OF_FILES)` to make sure that it is not empty before purging resumes

rishabh-d-dave replied:
I don't see the context of this review comment since quoted code is -

```
        time.sleep(2)

        # XXX actual test here: test that purging is resumed and finished
        self.config_set('mgr', self.conf_opt, False)
```

rishabh-d-dave replied:
Same for this one, the code has moved/removed since this review comment was posted.

rishabh-d-dave replied:
> I meant to say we should check that trash is not empty before setting the config option to False, then only it make sense to resume purging and deleting non-empty trash dir

The trash would surely empty. so it's not a worry.

rishabh-d-dave replied:
Marking this as resolved.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4337,298 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        o = o.lower().strip()
+        self.assertEqual(o, 'false')
+
+    def test_that_subvols_are_purged_by_default(self):
```
Isn't https://github.com/ceph/ceph/pull/54396/files#diff-d3a7e3f3f24fff510b4d2a562b2093257b20b3908748c3432d14e460c449186bR4435 an this test performing similar function?

rishabh-d-dave replied:
Right. Tests are working fine but some optimizations (like this one) can be made. I mentioned it here - https://github.com/ceph/ceph/pull/54396#issuecomment-2383071861.

rishabh-d-dave commented:
> > @vshankar Good to proceed to add code for pausing cloner thread?
> 
> Given the nature of changes, I'd say let's keep that as a separate change (and probably have someone else give a shot at implementing it).

@vshankar 
We had decided that we want to get it done in this same PR and therefore I've already worked on this. That's core reason behind current arrangement of commits on this PR branch.

vshankar commented:
> @vshankar
> We had decided that we want to get it done in this same PR and therefore I've already worked on this. That's core reason behind current arrangement of commits on this PR branch.

Yes. I'm fine with adding commits to this change for that, but I feel this should be put to test soon and is pretty close to merge.

rishabh-d-dave commented:
> BTW, @rishabh-d-dave do you already have the commits of disabling cloner threads. If you have already worked on that, please update this PR with those commits.

On it

rishabh-d-dave commented:
Rebased to resolve the conflict.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4354,407 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        o = o.lower().strip()
+        self.assertEqual(o, 'false')
+
+    def test_when_true_cloning_never_begins(self):
```
descriptions missing in the tests cases. test name doesn't tell which config is true ?

rishabh-d-dave replied:
> test name doesn't tell which config is true

It doesn't need to, class name does that. All tests under `TestPausePurging` are to `pause_purging` and all tests under `TestPauseCloning` are for `pause_cloninig`.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4354,407 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+
+        super().tearDown()
+
+    def test_the_default_value(self):
```
ditto, as for `TestPausePurging`

rishabh-d-dave replied:
The chance of default value being changed is low but having these test doesn't hurt and they'll catch if some bug gets introduced by accident.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4354,407 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        self.config_set('mgr', self.conf_opt, False)
+        self._wait_for_trash_empty()
+
+    def test_the_default_value(self):
```
Is this really required as a test case?

rishabh-d-dave replied:
The chance of default value being changed is low but having these test doesn't hurt and they'll catch if some bug gets introduced by accident.

neesingh-rh commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4354,407 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+            for filename in sv_files:
+                self.assertIn(filename, sv_files_new)
+
+    def test_when_true_subvol_is_trashed_but_stays_unpurged(self):
```
Similarly, we can have like:`test_subvolume_when_pause_purging_trash_is_enabled`. The goal of the test case is anyways mentioned in the description

rishabh-d-dave replied:
In logs test names appears like this: `tasks.cephfs.test_volumes.TestPausePurging.test_when_true_subvol_is_trashed_but_stays_unpurged`, so adding config option name is test name again is not needed because test class already mentioned it.

rishabh-d-dave commented:
Resolved the conflict.

rishabh-d-dave commented:
pinging once more for reviews

neesingh-rh commented:
jenkins test api

neesingh-rh reviewed (commented):
@rishabh-d-dave There are many older comments which are still un-addressed/not ACKED?

rishabh-d-dave commented:
> @rishabh-d-dave There are many older comments which are still un-addressed/not ACKED?

Older comments have been addressed or has been replied to.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
> The PR looks almost good, apart from many unaddressed comments (some review comments are of @vshankar )

They've been addressed, comments were not just marked resolved.

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -333,6 +333,15 @@ def _wait_for_trash_empty(self, timeout=60):
         # [sub]volume interface for this).
         trashdir = os.path.join("./", "volumes", "_deleting")
+
+        try:
```
Use `mount_a.stat()`.

rishabh-d-dave replied:
Done.

vshankar commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -550,7 +550,12 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
+            'pause_cloning',
+            type='bool',
+            default=False,
+            desc='Pause the threads that asynchronously clone snapshots')
```
nit: Pause asynchronous cloner threads.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        time.sleep(7)
+        self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
+
+    def test_when_changed_to_false_unpurged_subvol_is_purged(self):
```
There are two cases when purge threads are resumed: one, purging takes place when a new subvol is trashed, and two, purging for a partly purged subvol continues (and finishes). This name is ambiguous, it doesn't tell which case this test method is covering.

rishabh-d-dave replied:
I've modified all test names to a better version, LMK if that looks better.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -192,6 +204,47 @@ def run(self):
+
+    def resume(self):
+        if self.run_event.is_set():
+            log.info('resume() no need to resume, run_event is already set.')
```
IMO it'll be useful to log at INFO level that threads are resumed or purged.

rishabh-d-dave replied:
> IMO it'll be useful to log at INFO level that threads are resumed or purged.

@vshankar ping

vshankar replied:
What this ping for? There wasn't a question to be answered.

rishabh-d-dave replied:
So you agree with it? Our views about log level are opposing.

vshankar replied:
This is confusing. You did mention than the change has been made for now. I will revisit when I review the change again.

rishabh-d-dave replied:
I've made the change for now because you have asked me to make it but I think info level is better for these log messages.

rishabh-d-dave replied:
Had a chat with Venky about this, we'll prefer to this on debug level.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -192,6 +204,47 @@ def run(self):
+        # Therefore, set jobthread.cancel_event for other threads as well.
+        log.info('pause() pausing rest of worker threads')
+        for t in self.threads:
+            if not t.cancel_event.is_set():
```
Yes, that's why the big note is added.

vshankar replied:
I _did_ read that comment and I still couldn't infer why is this pre-check required.

rishabh-d-dave replied:
IIUC, you meant does it matter to check if `cancel_event` is set _before we set it on next line_, right? It doesn't really but I didn't want to call `threading.Event.set()` again when it already is set.

vshankar replied:
> IIUC, you meant does it matter to check if `cancel_event` is set _before we set it on next line_, right? It doesn't really but I didn't want to call `threading.Event.set()` again when it already is set.

Right, that's what I meant. And yes, setting it again isn't going have any negative consequences.

rishabh-d-dave replied:
Yeah, true. It was natural to write that "set threads that aren't set by `cancel_all_jobs()`."

vshankar replied:
OK. I see you agreed to this but I don't see the updated code.

rishabh-d-dave replied:
As per our last discussion, I've added a comment on top of this code.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+    def setUp(self):
+        super().setUp()
+
+        # so that we don't see log messages meant for clones. makes it easy
```
Removed.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        changed to false, the async purging of a subvolume will resume and
+        also finish, causing trash to be empty.
+        '''
+        v = self.volname
```
Following convention from previous tests, nothing otherwise.

rishabh-d-dave commented on PendingReleaseNotes:
```diff
@@ -28,6 +28,9 @@
 * MGR: MGR's always-on modulues/plugins can now be force-disabled. This can be
   necessary in cases where MGR(s) needs to be prevented from being flooded by
   the module commands when coresponding Ceph service is down/degraded.
+* CephFS: It is now possible to pause the threads that asynchronously purge
```
For now, I've squashed commits for both doc changes into one and both release note additions into one.

rishabh-d-dave commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -959,6 +959,13 @@ Configure the maximum number of concurrent clone operations. The default is 4:
+
+.. prompt:: bash #
+
+    ceph config set mgr/volumes/pause_purging
```
Fixed.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+
+        self.assertEqual(len(sv_files),  num_of_files)
+
+    def test_when_changed_to_true_ongoing_purge_halts(self):
```
I don't find this name comprehensible.

rishabh-d-dave replied:
I've modified all test names to a better version, LMK if that looks better.

vshankar replied:
> I don't find this name comprehensible.

Using `true/false` in test name is far more incomprehensible.

vshankar replied:
If that's the argument then use set/unset instead of true/false.

I will have a look at the updated test names.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        self._assert_trashed_sv_has_num_of_files(trashed_sv_path,
+                                                 NUM_OF_FILES)
+
+    def test_when_changed_to_false_partly_purged_subvol_purges_fully(self):
```
I've modified all test names to a better version, LMK if that looks better.

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        for filename in sv_files:
+            self.assertNotIn(filename, sv_files_new)
+
+    def _assert_trashed_sv_is_unpurged(self, sv, sv_path, sv_files):
```
I didn't opt for that since we already had subvol path, so better to pass it.

vshankar replied:
Ok, will check when I revisit the change again.

vshankar replied:
But you need to construct the trash dir, so you just need the uuid (which you fetch) and then construct the path, isn't it?

rishabh-d-dave replied:
Hey, Venky. I checked this, it's simpler this way. When we have a path like this `/tmp/tmphyzpnx6r/mnt.0/volumes/_nogroup/sv1/4824a595-9f5f-4af1-ab04-8f1ec7a624ea`, we need the subvol path (to get the UUID dir), the subvol name (for easy removal of it) and subvol files (to build complete path to files in trash).

vshankar replied:
Ok, I'll buy that argument 👍

rishabh-d-dave commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+            for filename in sv_files:
+                self.assertIn(filename, sv_files_new)
+
+    def test_when_true_subvol_is_trashed_but_stays_unpurged(self):
```
There are two cases here too (both covered in current set of tests). This name is ambiguous since it doesn't tell which case we are trying to cover.

rishabh-d-dave replied:
Marking this and other similar comment resolved since test name has been changed and Venky intends to go through this PR again.

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        self.run_ceph_cmd(f'fs subvolume rm {v} {sv}')
+        self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
+        time.sleep(7)
+        self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
```
I understand that. My point is you already know the number of files and checking that none of those files have been purged by waiting for 60 seconds is a good enough check.

rishabh-d-dave replied:
Waiting is redundant if files were deleted at the very beginning of waiting period.

About the point about number of files, I agree.

vshankar replied:
How is it redundant. The config is set before removing the subvolume, so, not a single file should get purged.

rishabh-d-dave replied:
Ideally, but not in case of bug.

vshankar replied:
In that case, would checking twice ensure that the bug would be caught?

vshankar replied:
This is not making any sense @rishabh-d-dave 

So, you mean:

```
check_something()
sleep(...)
check_something()
```

is better than

```
wait(...)
check_something()
```

especially, when `check_something()` would know the count of the number of files to compare with, just because the former would catch bugs early?

rishabh-d-dave replied:
If `subvolume rm` will move subvolume to trash and _if we have a bug_, purging would begin immediately even though `pause_purging` is set. In such a case, waiting/sleeping is not needed. `check_something()` (the one before sleep/wait) would fail and cause test to abort.

vshankar replied:
@rishabh-d-dave It seems like you are trying to save a minute or so in test runs, yes? I would prefer simple/straight-froward approaches (especially in test code) in lieu of saving _small_ runtime.

rishabh-d-dave replied:
Made the change.

vshankar commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -544,7 +544,13 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             type='bool',
             default=True,
-            desc='Reject subvolume clone request when cloner threads are busy')
+            desc='Reject subvolume clone request when cloner threads are busy'),
```
Why? This commit isn't a huge change and all it does it invoke the respective pause/resume calls.

rishabh-d-dave replied:
> Why? This commit isn't a huge change and all it does it invoke the respective pause/resume calls.

It's not about size, it's about the fact that there are 3 logical set of source code modifications: 1, code to halt async job machinery, 2, code to pause/resume purge threads, and 3, code to pause/resume clone threads.

Keeping these logically different set of changes in separate commits not only allows documenting them properly but also makes it easier to understand, which will not be the case when they are lumped together.

vshankar replied:
The bulk of the change is the async job machinery. So you could club the purge and clone part that just invokes the purge/resume call in the async machinery.

rishabh-d-dave replied:
Discussed this in a meeting with Venky, squashing all source code changes into one.

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+
+        self.run_ceph_cmd(f'fs subvolume rm {v} {sv}')
+        self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
+        time.sleep(7)
```
which is magic anyway.

vshankar replied:
ping?? (unresolved comment)

vshankar replied:
use safe_while where ever possible.

rishabh-d-dave replied:
The drawback of this is that it looks unnecessary longer and complex -

```
with safe_while(tries=1, sleep=7):
    try:
        while proceed():
            self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
     except MaxWhileTries:
        pass
```

This is how code originally looks -

```
time.sleep(7)
self._assert_trashed_sv_is_unpurged(sv, sv_path, sv_files)
```

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4352,317 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        if len(sv_files) == len(sv_files_new):
+            return
+        else:
+            for filename in sv_files:
```
This is reached when the number of files don't match. Why doesn't the test fail right away?

rishabh-d-dave replied:
Pointless to run the loop if number of files before and after aren't the same.

rishabh-d-dave replied:
I also mentioned the condition under which it is pointless in the same sentence.

vshankar replied:
Sorry, you are confusing me @rishabh-d-dave 

First of all, there is return under the if condition and thereby the else block is not necessary. Secondly, if the number of files mismatch, the loop checks if the passed in file names are in the scanned list of files. I didn't really get what is being done and I went back in to the previous function (L#4400) which says:

> # in case dir for subvol is undeleted yet (but files inside it are).

So, if you had 10 files in `sv_files` and you get 8 files by scanning (i.e., in `sv_files_new`), then assert is surely going to be hit, isn't it?

vshankar replied:
If the whole point of that loop is to fail the test, then there are better, non-cryptic ways to doing so.

rishabh-d-dave replied:
Since as per our chat on Slack the name of missing doesn't really help and since we don't want any optimization, I am replacing this entire by the for loop it contains.

rishabh-d-dave commented:
@vshankar Leaving gentle reminder once more, please review when you find some time.

vshankar commented:
> @vshankar Leaving gentle reminder once more, please review when you find some time.

Imy on pto till tomorrow post cephalocon.

rishabh-d-dave commented:
> > @vshankar Leaving gentle reminder once more, please review when you find some time.
> 
> Imy on pto till tomorrow post cephalocon.

I am aware of it, just leaving a reminder in advance :)

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
@vshankar pinging again

vshankar commented:
> @vshankar pinging again

On this today.

rishabh-d-dave commented:
@vshankar pinging once more, a gentle reminder  :)

vshankar commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -4345,6 +4345,190 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
+        Test that when MGR config option mgr/volumes/pause_purging is
+        is set to false, trashed but unpurged subvolume is purged (fully).
+        '''
+        v = self.volname
```
marked as resolved without an explanation??

vshankar replied:
OK, I see this was updated in a later comment that test convention was followed. nvm.

rishabh-d-dave commented:
@vshankar PTAL

dparmar18 commented on src/pybind/mgr/volumes/fs/async_job.py:
```diff
@@ -192,6 +204,47 @@ def run(self):
+            t.cancel_event.clear()
+
+        self.run_event.set()
+        log.debug('resume() run_event has been set, queue and threads have been'
+                  'resumed')
```
extreme nit: need a space before `resumed` or after `been`

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/69939](https://tracker.ceph.com/issues/69939).

vshankar reviewed (commented):
I'm seeing a couple of test failures related to volume plugin. This change is the only mgr/volumes related changes in the test branch.

- https://pulpito.ceph.com/vshankar-2025-02-27_13:15:04-fs-wip-vshankar-testing-20250227.075314-debug-testing-default-smithi/8158600/
- https://pulpito.ceph.com/vshankar-2025-02-27_13:15:04-fs-wip-vshankar-testing-20250227.075314-debug-testing-default-smithi/8158621/

@rishabh-d-dave - could you please check the failures to assess if those are related to this change? If not, I will create trackers for the same and proceed with merging.

rishabh-d-dave commented:
@vshankar Issue has been fixed. All tests in fs:volumes are passing now - https://pulpito.ceph.com/rishabh-2025-03-10_02:36:23-fs:volumes-rishabh-mgr-vol-no-del-testing-default-smithi/

rishabh-d-dave commented:
about last push: removed DNM commit

vshankar commented:
> @vshankar Issue has been fixed. All tests in fs:volumes are passing now - https://pulpito.ceph.com/rishabh-2025-03-10_02:36:23-fs:volumes-rishabh-mgr-vol-no-del-testing-default-smithi/

Could you explain what the issue was and the commit (new or existing) where the fix has been made?

rishabh-d-dave commented:
> > @vshankar Issue has been fixed. All tests in fs:volumes are passing now - https://pulpito.ceph.com/rishabh-2025-03-10_02:36:23-fs:volumes-rishabh-mgr-vol-no-del-testing-default-smithi/
> 
> Could you explain what the issue was and the commit (new or existing) where the fix has been made?

The fix is in the last diff, specifically here - https://github.com/ceph/ceph/compare/db7d9131f4e2d70ec627d5166f76891714d60306..69c8ec7d06011e4632c55872d7d356fdf34038d5#diff-7ac80b107445f14d5cfc30196533c08eaafd56c101cae84a0f55df526a229eceR309. It was due to a leftover code when the name of the variable was changed.

rishabh-d-dave commented:
jenkins test make check

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/70394](https://tracker.ceph.com/issues/70394).

vshankar commented:
@rishabh-d-dave please rebase.

rishabh-d-dave commented:
jenkins test make check

vshankar commented:
fs:volumes run: https://pulpito.ceph.com/vshankar-2025-03-13_10:45:26-fs:volumes-wip-vshankar-testing-20250313.072951-debug-testing-default-smithi/

I'm running the entire fs suite too, but for this change fs:volumes test would suffice and can be merged once that passes. I will link the run wiki entry later.
